File: static/g2048_engine.js
/*
 * Deterministic 2048 engine — shared by the browser (gameplay) and the server
 * (replay). 2048's only randomness is tile spawning: which empty cell, and
 * 2 (90%) vs 4 (10%). Both are driven by the same integer mulberry32 PRNG used
 * by the other engines, consumed in a fixed order (cell first, then value), so
 * JS and Python evolve identically from the same seed + move sequence and the
 * server can recompute the authoritative score from the recorded moves.
 *
 * Inputs are direction presses recorded as { tick, dir }, dir in:
 *   U = up, D = down, L = left, R = right.
 * A move that does not change the board is a no-op (no spawn, no RNG, the move
 * counter does not advance) — exactly as in real 2048. The game ends when the
 * board is full and no merge is possible.
 *
 * UMD wrapper: browser global `G2048Engine`, or require() under Node.
 */
(function (root, factory) {
  const api = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else root.G2048Engine = api;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  "use strict";

  const SIZE = 4;
  const TARGET = 2048;
  const MAX_MOVES = 100000;
  const ACTIONS = ["U", "D", "L", "R"];

  function makeRng(seed) {
    let a = seed >>> 0;
    return function () {
      a |= 0;
      a = (a + 0x6d2b79f5) | 0;
      let t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = ((t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t) >>> 0;
      return (t ^ (t >>> 14)) >>> 0;
    };
  }

  // For each direction, the four lanes (board indices, row-major i = r*4 + c)
  // ordered FROM the edge tiles slide toward; compressing each lane toward its
  // index 0 performs the move.
  function lanesFor(direction) {
    const lanes = [];
    if (direction === "L") {
      for (let r = 0; r < SIZE; r++) {
        const lane = [];
        for (let c = 0; c < SIZE; c++) lane.push(r * SIZE + c);
        lanes.push(lane);
      }
    } else if (direction === "R") {
      for (let r = 0; r < SIZE; r++) {
        const lane = [];
        for (let c = SIZE - 1; c >= 0; c--) lane.push(r * SIZE + c);
        lanes.push(lane);
      }
    } else if (direction === "U") {
      for (let c = 0; c < SIZE; c++) {
        const lane = [];
        for (let r = 0; r < SIZE; r++) lane.push(r * SIZE + c);
        lanes.push(lane);
      }
    } else if (direction === "D") {
      for (let c = 0; c < SIZE; c++) {
        const lane = [];
        for (let r = SIZE - 1; r >= 0; r--) lane.push(r * SIZE + c);
        lanes.push(lane);
      }
    }
    return lanes;
  }

  const LANES = {};
  for (const d of ACTIONS) LANES[d] = lanesFor(d);

  // Slide+merge a 4-length lane toward index 0. Returns { out, gained, changed }.
  function compress(vals) {
    const tiles = [];
    for (let i = 0; i < vals.length; i++) if (vals[i] !== 0) tiles.push(vals[i]);
    const out = [];
    let gained = 0;
    let i = 0;
    const n = tiles.length;
    while (i < n) {
      if (i + 1 < n && tiles[i] === tiles[i + 1]) {
        const merged = tiles[i] * 2;
        out.push(merged);
        gained += merged;
        i += 2;
      } else {
        out.push(tiles[i]);
        i += 1;
      }
    }
    while (out.length < SIZE) out.push(0);
    let changed = false;
    for (let k = 0; k < SIZE; k++) if (out[k] !== vals[k]) { changed = true; break; }
    return { out: out, gained: gained, changed: changed };
  }

  function boardMax(board) {
    let m = 0;
    for (let i = 0; i < board.length; i++) if (board[i] > m) m = board[i];
    return m;
  }

  class Engine {
    constructor(seed) {
      this.size = SIZE;
      this.rng = makeRng(seed >>> 0);
      this.board = new Array(SIZE * SIZE).fill(0);
      this.score = 0;
      this.moves = 0;
      this.highest = 0;
      this.won = false;
      this.gameOver = false;
      this.lastSpawn = -1; // cell index of the most recent spawn (display only)
      this._spawn();
      this._spawn();
      this._updateStatus();
    }

    _spawn() {
      const empties = [];
      for (let i = 0; i < this.board.length; i++) if (this.board[i] === 0) empties.push(i);
      if (empties.length === 0) { this.lastSpawn = -1; return; }
      const cell = empties[this.rng() % empties.length];
      const value = (this.rng() % 10) === 0 ? 4 : 2;
      this.board[cell] = value;
      this.lastSpawn = cell;
    }

    _updateStatus() {
      const m = boardMax(this.board);
      if (m > this.highest) this.highest = m;
      if (m >= TARGET) this.won = true;
      if (this.board.indexOf(0) !== -1) { this.gameOver = false; return; }
      let movable = false;
      for (let r = 0; r < SIZE; r++) {
        for (let c = 0; c < SIZE; c++) {
          const v = this.board[r * SIZE + c];
          if (c + 1 < SIZE && this.board[r * SIZE + c + 1] === v) movable = true;
          if (r + 1 < SIZE && this.board[(r + 1) * SIZE + c] === v) movable = true;
        }
      }
      this.gameOver = !movable;
    }

    move(direction) {
      if (this.gameOver || !LANES[direction]) return false;
      const work = this.board.slice();
      let changed = false;
      let gained = 0;
      const lanes = LANES[direction];
      for (let li = 0; li < lanes.length; li++) {
        const lane = lanes[li];
        const vals = [work[lane[0]], work[lane[1]], work[lane[2]], work[lane[3]]];
        const res = compress(vals);
        if (res.changed) changed = true;
        gained += res.gained;
        for (let k = 0; k < SIZE; k++) work[lane[k]] = res.out[k];
      }
      if (!changed) return false;
      this.board = work;
      this.score += gained;
      this.moves += 1;
      this._spawn();
      this._updateStatus();
      return true;
    }
  }

  function simulate(seed, inputs, maxMoves) {
    maxMoves = maxMoves || MAX_MOVES;
    const e = new Engine(seed);
    let count = 0;
    for (const it of inputs) {
      if (e.gameOver || count >= maxMoves) break;
      const d = it && it.dir;
      if (ACTIONS.indexOf(d) === -1) continue;
      e.move(d);
      count += 1;
    }
    return {
      score: e.score, highest: e.highest, moves: e.moves,
      won: e.won, gameOver: e.gameOver,
    };
  }

  return { Engine, simulate, makeRng, ACTIONS, SIZE, TARGET };
});

File: g2048_engine.py
"""
Python port of static/g2048_engine.js — must produce bit-identical results so
the server can replay a recorded 2048 game and compute the authoritative score.

2048 has exactly one source of randomness: which empty cell a new tile spawns
in, and whether it is a 2 (90%) or a 4 (10%). Both are driven by the same
seedable mulberry32 PRNG used by the other engines, consumed in a fixed order
(first the cell, then the value), so JS and Python evolve identically from the
same seed and the same move sequence.

Inputs are recorded as { tick, dir }, dir in:  U = up, D = down, L = left,
R = right.  A move that does not change the board is a no-op: it neither
advances the move counter nor consumes RNG (no tile spawns) — exactly as in
real 2048. The game ends when the board is full and no merge is possible.
"""

SIZE = 4
TARGET = 2048                # tile value that flips the "won" flag (play continues)
MAX_MOVES = 100_000          # safety bound on replay length
ACTIONS = ("U", "D", "L", "R")

_MASK = 0xFFFFFFFF


def _u32(x):
    return x & _MASK


def _i32(x):
    x &= _MASK
    return x - 0x100000000 if x & 0x80000000 else x


def _imul(a, b):
    return _i32((_i32(a) * _i32(b)) & _MASK)


def _ushr(x, n):
    return _u32(x) >> n


def _xor(x, y):
    return _i32(_i32(x) ^ _i32(y))


def _or(x, y):
    return _i32(_i32(x) | _i32(y))


class _Rng:
    """mulberry32 — identical 32-bit arithmetic to makeRng() in the JS engine."""

    def __init__(self, seed):
        self.a = _i32(seed)

    def next(self):
        a = _i32(self.a)
        a = _i32(a + 0x6D2B79F5)
        self.a = a
        t = _imul(_xor(a, _ushr(a, 15)), _or(1, a))
        t = _u32(_xor(_i32(t + _imul(_xor(t, _ushr(t, 7)), _or(61, t))), t))
        return _u32(_xor(t, _ushr(t, 14)))


# For each direction, the four "lanes" (lists of board indices, row-major i=r*4+c)
# ordered FROM the edge tiles slide toward. Compressing each lane toward its
# index 0 therefore performs the move for that direction.
def _lanes(direction):
    lanes = []
    if direction == "L":
        for r in range(SIZE):
            lanes.append([r * SIZE + c for c in range(SIZE)])
    elif direction == "R":
        for r in range(SIZE):
            lanes.append([r * SIZE + c for c in range(SIZE - 1, -1, -1)])
    elif direction == "U":
        for c in range(SIZE):
            lanes.append([r * SIZE + c for r in range(SIZE)])
    elif direction == "D":
        for c in range(SIZE):
            lanes.append([r * SIZE + c for r in range(SIZE - 1, -1, -1)])
    return lanes


_LANES = {d: _lanes(d) for d in ACTIONS}


def _compress(vals):
    """Slide+merge a 4-length lane toward index 0. Returns (out, gained, changed)."""
    tiles = [v for v in vals if v != 0]
    out = []
    gained = 0
    i = 0
    n = len(tiles)
    while i < n:
        if i + 1 < n and tiles[i] == tiles[i + 1]:
            merged = tiles[i] * 2
            out.append(merged)
            gained += merged
            i += 2
        else:
            out.append(tiles[i])
            i += 1
    while len(out) < SIZE:
        out.append(0)
    return out, gained, out != vals


class Engine:
    def __init__(self, seed):
        self.size = SIZE
        self.rng = _Rng(seed)
        self.board = [0] * (SIZE * SIZE)
        self.score = 0
        self.moves = 0
        self.highest = 0
        self.won = False
        self.game_over = False
        self.last_spawn = -1          # cell index of the most recent spawn (display only)
        self._spawn()
        self._spawn()
        self._update_status()

    def _spawn(self):
        empties = [i for i in range(SIZE * SIZE) if self.board[i] == 0]
        if not empties:
            self.last_spawn = -1
            return
        cell = empties[self.rng.next() % len(empties)]
        value = 4 if (self.rng.next() % 10) == 0 else 2
        self.board[cell] = value
        self.last_spawn = cell

    def _update_status(self):
        m = max(self.board)
        if m > self.highest:
            self.highest = m
        if m >= TARGET:
            self.won = True
        if 0 in self.board:
            self.game_over = False
            return
        # Board full: game over unless some neighbour pair is equal.
        movable = False
        for r in range(SIZE):
            for c in range(SIZE):
                v = self.board[r * SIZE + c]
                if c + 1 < SIZE and self.board[r * SIZE + c + 1] == v:
                    movable = True
                if r + 1 < SIZE and self.board[(r + 1) * SIZE + c] == v:
                    movable = True
        self.game_over = not movable

    def move(self, direction):
        if self.game_over or direction not in _LANES:
            return False
        work = self.board[:]
        changed = False
        gained = 0
        for lane in _LANES[direction]:
            out, g, ch = _compress([work[i] for i in lane])
            if ch:
                changed = True
            gained += g
            for k in range(SIZE):
                work[lane[k]] = out[k]
        if not changed:
            return False
        self.board = work
        self.score += gained
        self.moves += 1
        self._spawn()
        self._update_status()
        return True


def simulate(seed, inputs, max_moves=MAX_MOVES):
    """Replay a recorded game. inputs: iterable of {"tick": int, "dir": str}."""
    e = Engine(seed)
    count = 0
    for it in inputs:
        if e.game_over or count >= max_moves:
            break
        try:
            d = it["dir"]
        except (KeyError, TypeError):
            continue
        if d not in ACTIONS:
            continue
        e.move(d)
        count += 1
    return {
        "score": e.score,
        "highest": e.highest,
        "moves": e.moves,
        "won": e.won,
        "gameOver": e.game_over,
    }
